File: bindings/python/aster/protocol.py
"""
aster.protocol -- Wire-protocol dataclasses.

Spec reference: §6.2 (StreamHeader), §6.4 (RpcStatus/trailer)

These types are always serialized with Fory XLANG, regardless of the
service's negotiated serialization mode. Every field uses
``pyfory.field(id=N)`` so the Fory struct-fingerprint is tag-ID based and
stable across bindings -- without explicit IDs Java's Fory would snake_case
the field name when computing its fingerprint while Python leaves it
verbatim, and the two sides hash-mismatch with
``Hash X is not consistent with Y for type T`` at decode time. IDs here
MUST stay in sync with the ``@ForyField(id = N)`` annotations on the
matching Java types under
``bindings/java/aster-runtime/src/main/java/site/aster/server/wire/``.
"""

from __future__ import annotations

from dataclasses import dataclass

import pyfory

# Import wire_type from codec.py to avoid duplication
from aster.codec import wire_type


# ── Framework-internal protocol types ────────────────────────────────────────


@dataclass
@wire_type("_aster/StreamHeader")
class StreamHeader:
    """First frame on every QUIC stream (HEADER flag).

    Carries service routing, contract identity, call metadata, and
    the negotiated serialization mode.
    """

    service: str = pyfory.field(0, default="")
    method: str = pyfory.field(1, default="")
    version: pyfory.int32 = pyfory.field(2, default=0)
    callId: pyfory.int32 = pyfory.field(3, default=0)
    # relative seconds, 0 = no deadline
    deadline: pyfory.int16 = pyfory.field(4, default=0)
    # XLANG=0, NATIVE=1, ROW=2, JSON=3
    serializationMode: pyfory.int8 = pyfory.field(5, default=0)
    metadataKeys: list[str] = pyfory.field(6, default_factory=list)
    metadataValues: list[str] = pyfory.field(7, default_factory=list)
    # Session identifier (multiplexed-streams, spec §6). 0 = stateless
    # SHARED pool stream; non-zero = stream belongs to the session with
    # this id on this (peer, connection). Monotonically allocated
    # client-side per connection.
    sessionId: pyfory.int32 = pyfory.field(8, default=0)


@dataclass
@wire_type("_aster/CallHeader")
class CallHeader:
    """Per-call header within a session stream (CALL flag).

    Used for session-scoped services (Phase 8) where multiple RPCs
    share a single QUIC stream.
    """

    method: str = pyfory.field(0, default="")
    callId: pyfory.int32 = pyfory.field(1, default=0)
    # relative seconds, 0 = no deadline
    deadline: pyfory.int16 = pyfory.field(2, default=0)
    metadataKeys: list[str] = pyfory.field(3, default_factory=list)
    metadataValues: list[str] = pyfory.field(4, default_factory=list)


@dataclass
@wire_type("_aster/RpcStatus")
class RpcStatus:
    """Trailing status frame (TRAILER flag).

    Sent as the last frame on a stream to communicate the outcome of
    the RPC to the peer.
    """

    code: pyfory.int32 = pyfory.field(0, default=0)
    message: str = pyfory.field(1, default="")
    detailKeys: list[str] = pyfory.field(2, default_factory=list)
    detailValues: list[str] = pyfory.field(3, default_factory=list)

File: bindings/typescript/packages/aster/src/protocol.ts
/**
 * Wire-protocol types.
 *
 * Spec reference: S6.2 (StreamHeader), S6.4 (RpcStatus/trailer)
 *
 * These types are always serialized with Fory XLANG, regardless of the
 * service's negotiated serialization mode.
 */

/** First frame on every QUIC stream (HEADER flag). */
export class StreamHeader {
  /** Wire type tag for Fory XLANG registration. */
  static readonly wireType = '_aster/StreamHeader';

  service = '';
  method = '';
  version = 0;
  callId = 0;               // int32 sequence number
  deadline = 0;             // int16 relative seconds, 0 = none
  serializationMode = 0;    // int8: XLANG=0, NATIVE=1, ROW=2, JSON=3
  metadataKeys: string[] = [];
  metadataValues: string[] = [];
  // Session identifier (multiplexed-streams, spec §6). 0 = stateless
  // SHARED pool stream; non-zero = stream belongs to the session with
  // this id on this (peer, connection). Client allocates monotonically
  // per connection. On the wire this is an int32, matching pyfory.
  sessionId = 0;

  constructor(init?: Partial<StreamHeader>) {
    if (init) Object.assign(this, init);
  }
}

/** Per-call header within a session stream (CALL flag). */
export class CallHeader {
  static readonly wireType = '_aster/CallHeader';

  method = '';
  callId = 0;               // int32 sequence number
  deadline = 0;             // int16 relative seconds, 0 = none
  metadataKeys: string[] = [];
  metadataValues: string[] = [];

  constructor(init?: Partial<CallHeader>) {
    if (init) Object.assign(this, init);
  }
}

/** Trailing status frame (TRAILER flag). */
export class RpcStatus {
  static readonly wireType = '_aster/RpcStatus';

  code = 0;
  message = '';
  detailKeys: string[] = [];
  detailValues: string[] = [];

  constructor(init?: Partial<RpcStatus>) {
    if (init) Object.assign(this, init);
  }

  /** Convert detail arrays to a key-value record. */
  get details(): Record<string, string> {
    const result: Record<string, string> = {};
    for (let i = 0; i < this.detailKeys.length; i++) {
      result[this.detailKeys[i]!] = this.detailValues[i] ?? '';
    }
    return result;
  }
}
